### PR TITLE
Add property, method, and indexed item splatting support

### PR DESCRIPTION
**Pull request by KirkMunro**
# PR Summary

Like PR's #10984 and #10073, this PR enhances the splatting capabilities in PowerShell.

With this PR in place, users are able to splat members (properties and methods) as well as indexed items into an invocation. It doesn't matter how many levels deep you go with your expression.

Examples:

```powershell
# Splat a property
Get-Date @data.Parameters

# Splat the results of a method
Get-Date @data.GetParams()

# Splat a nested property
Get-Process @PSCmdlet.MyInvocation.BoundParameters

# Splat a property from the result of a method
Test-Splat @result.GetParameterSets().Value

# Splat an item from a collection
Test-Splat @parameterSet[0]
```

The only question I have is whether or not this functionality needs to be behind an experimental feature. If you review the code changes, you'll see that adding this support was very, very easy, and I believe the changes are low risk, so I don't know that it is worth the added complexity of wrapping these in an experimental feature.

## PR Context

This functionality has already been identified as desirable for PowerShell as part of [RFC0002](https://github.com/PowerShell/PowerShell-RFC/blob/e09b14fe01b636ea38260da9e70ff024dc32936b/2-Draft-Accepted/RFC0002-Generalized-Splatting.md).

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/Compiler.cs [hunk 1]
```diff
@@ -4028,14 +4028,31 @@ public object VisitCommand(CommandAst commandAst)
                     var splatTest = element;
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: @KirkMunro, your last commit had 1 failures in `PowerShell-CI-windows`
Enter-PSHostProcess tests.By Process Id.Can enter using NamedPipeConnectionInfo
```powershell
Exception calling "Invoke" with "0" argument(s): "The runspace state is not valid for this operation."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Core\Enter-PSHostProcess.Tests.ps1: line 136
136:                 $ps.AddScript('$pid').Invoke() | Should -Be $pwshId

```

Comment (KirkMunro) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: @PoshChan please restart Windows

**Comment by KirkMunro** [comment]
@PoshChan please restart windows

**Comment by PoshChan** [comment]
@KirkMunro, successfully started retry of `PowerShell-CI-Windows`

File: test/powershell/Language/Scripting/Splatting.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                    [ValidateNotNullOrEmpty()]
+                    [string]
+                    $Value2
+                )
+                $PSCmdlet.ParameterSetName
+            }
+        }
+
+        It 'Should splat property values properly' {
+            $o = [pscustomobject]@{
```
Comment (iSazonov) [inline test/powershell/Language/Scripting/Splatting.Tests.ps1]: $o - please use more readable name.

File: test/powershell/Language/Scripting/Splatting.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                }
+                @{
+                    InputObject = 'Second'
+                }
+            )
+            Add-Member -InputObject $o -Name GetParams -MemberType ScriptMethod -Value { $this.Parameters[$args[0]] }
+            1..1 | ForEach-Object -Parallel { Write-Output @using:o[1] } | Should -BeExactly 'Second'
+        }
+
+        It 'Should splat with properties and indexed data and using: properly' {
```
Comment (iSazonov) [inline test/powershell/Language/Scripting/Splatting.Tests.ps1]:
```suggestion
        It 'Should splat with properties and indexed data and 'using:' properly' {
```

File: test/powershell/Language/Scripting/Splatting.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        }
+
+        AfterAll {
+            if ($skipTest) {
+                $global:PSDefaultParameterValues = $originalDefaultParameterValues
+            }
+        }
+
+        It 'Should splat indexed data and using: properly' {
```
Comment (iSazonov) [inline test/powershell/Language/Scripting/Splatting.Tests.ps1]:
```suggestion
        It 'Should splat indexed data and 'using:' properly' {
```

**Review by iSazonov — COMMENTED** [review]
I think it makes sense to add more tests that splatting (old and new) works (like Get-Date @data and Get-Date @data.Parameters )

File: test/powershell/Language/Scripting/Splatting.Tests.ps1 (hunk 2 above)
Comment (KirkMunro) [reply test/powershell/Language/Scripting/Splatting.Tests.ps1]: Fixed.

**Comment by KirkMunro** [comment]
> I think it makes sense to add more tests that splatting (old and new) works (like Get-Date @DaTa and Get-Date @data.Parameters )

Agreed. I added some basic splatting tests since we didn't have specific tests in place already.

**Comment by iSazonov** [comment]
@KirkMunro Please fix CIs failures.

**Comment by KirkMunro** [comment]
@PoshChan please retry all

**Comment by PoshChan** [comment]
@KirkMunro, successfully started retry of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

File: src/System.Management.Automation/engine/parser/Compiler.cs (hunk 1 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: @KirkMunro, your last commit had 1 failures in `PowerShell-CI-macos`
Tests for splatting.Splatting members and indexed data with the using statement.Error occurred in Context block
```powershell
Cannot process argument transformation on parameter 'Test'. Cannot convert the "using: properly" value of type "System.String" to type "System.Management.Automation.ScriptBlock".
At /Users/runner/runners/2.160.1/work/1/s/test/powershell/Language/Scripting/Splatting.Tests.ps1:148 char:44
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: @KirkMunro, your last commit had 1 failures in `PowerShell-CI-linux`
Tests for splatting.Splatting members and indexed data with the using statement.Error occurred in Context block
```powershell
Cannot process argument transformation on parameter 'Test'. Cannot convert the "using: properly" value of type "System.String" to type "System.Management.Automation.ScriptBlock".
At /home/vsts/work/1/s/test/powershell/Language/Scripting/Splatting.Tests.ps1:148 char:44
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: @KirkMunro, your last commit had 1 failures in `PowerShell-CI-windows`
Tests for splatting.Splatting members and indexed data with the using statement.Error occurred in Context block
```powershell
Cannot process argument transformation on parameter 'Test'. Cannot convert the "using: properly" value of type "System.String" to type "System.Management.Automation.ScriptBlock".
At D:\a\1\s\test\powershell\Language\Scripting\Splatting.Tests.ps1:148 char:44
```

**Comment by iSazonov** [comment]
Should we update RFC0002 for the syntax?

**Comment by KirkMunro** [comment]
@iSazonov Someone can if they want to. [I shared the syntax 3+ years ago, before that RFC was approved](https://github.com/PowerShell/PowerShell-RFC/issues/6#issuecomment-189517556), but @lzybkr didn't add it to the RFC at the time.

**Comment by joeyaiello** [comment]
@PowerShell/powershell-committee reviewed this, and while we don't see a technical risk or breaking change, we're having trouble coming up with a compelling, real-world scenario to justify adding it to the language. 

Especially considering `Test-Splat @PSBoundParameters` works for the 3rd case, are there some scripting conveniences that can actually be gained from the other 4 examples you gave? In all the cases we could think of, what you have seems less clear than if you were to simply assign the value to another variable.

**Comment by KirkMunro** [comment]
The 3rd case is exactly the scenario where this is needed, because it doesn't always work. For example:

```powershell
function Test-Splat {
    [CmdletBinding()]
    param(
        [parameter(mandatory)]
        [string]
        $Name
    )
    $sb = {Get-Service @PSBoundParameters}
    & $sb
}
Test-Splat -Name wuauserv
```

If you run that command, you'll get all services. Why? Because `$PSBoundParameters` variable is current scope only. So if you're using nested scopes or script blocks to reuse code within a function, and you want to splat in bound parameters, you cannot. You instead need to create an intermediate variable to store `$PSCmdlet.MyInvocation.BoundParameters` in a temporary variable, and then splat that variable. I've had to use a temporary variable to achieve this many, many times in scripts and I continue to have to do it today. This is the most common scenario that drove me to add this to the language.

As a best practice, I don't use `$PSBoundParameters` in scripts because it is only accurate in the current scope, and I don't like minor refactoring risks to introduce unexpected issues so I use `$PSCmdlet.MyInvocation.BoundParameters` everywhere because it gives me consistency no matter where I use it, so I can refactor with confidence.

Here is an example demonstrating what I would have to do instead if I wanted to make what I shared above work today:

```powershell
function Test-Splat {
    [CmdletBinding()]
    param(
        [parameter(mandatory)]
        [string]
        $Name
    )
    $sb = {
        $cmdletParams = $PSCmdlet.MyInvocation.BoundParameters
        Get-Service @cmdletParams
    }
    & $sb
}
Test-Splat -Name wuauserv
```

The use of intermediate variables here feels unnecessary and it's just extra code to maintain. I want to be able to accomplish the same like this:

```powershell
function Test-Splat {
    [CmdletBinding()]
    param(
        [parameter(mandatory)]
        [string]
        $Name
    )
    $sb = {
        Get-Service @PSCmdlet.MyInvocation.BoundParameters
    }
    & $sb
}
Test-Splat -Name wuauserv
```

The other scenarios are added because it makes much more sense to me syntactically to be able to use the new syntax, which is very clear in intent, than to add an overly complex splat-subexpression syntax such as `@$($PSCmdlet.MyInvocation.BoundParameters)` to the language, as was suggested in the RFC.

Further, and quite important (so make sure you read/get this), these changes open the door for something else that is needed with splatting. One thing missing from splatting is the ability to perform a partial hashtable splat. Splat in one or more parameters, and only if they are present as keys in the hashtable. Splat in all parameters in the hashtable except for one or more parameters. Those scenarios (manipulating hashtables to get a subset hashtable easily without having to write complicated code) have benefits that extend beyond splatting and are useful elsewhere in scripts. By extending the splatting syntax to support member invocation and indices, we're opening the door for hashtable manipulation to be added independently of splatting, as it should be, such that splatting can just support it once it is added (maybe for free, maybe with a minor addition, depending on the syntax used for hashtable manipulation). If this is not clear, let me know and I'll go into more detail.

This need, and the justification for the syntax shared here, is crystal clear to me, keeping splatting focused on exactly what it was designed to do (take an array or a hashtable and splat those values into an invocation), while leveraging other current functionality (accessing members, accessing an indexed item) and new functionality (hashtable manipulation) using the same syntax that you would use elsewhere.

PowerShell is object-oriented, with properties and methods, as well as indexable collections. When working with variables, we can access properties and methods and indexed items in collections with ease. Why should we have to use a different syntax to do so when splatting, or inject intermediate variables, when we can simply change a `$` before a variable name to a `@` and as a result be able to splat whatever that would return, in a more natural syntax that we're already familiar with, without having to use subexpressions where subexpressions are normally not required?

**Comment by dragonwolf83** [comment]
Funny enough, I ran into this just the other day.

With the WinSCP PSModule there are parameters for HostName, Port, SshPrivateKeyPath, etc. They all match perfectly for Splatting. However, when you have multiple servers, you have to namespace them.

Ideally, I would do this:
```PowerShell
$config = @{
  QA = @{
    Host = @{
      HostName = 'abc'
      Port = 22
    }
  Prod = @{
    Host = @{
      HostName = 'abc'
      Port = 22
    }
  }
}

New-WinSCPOption @config.QA.Host
New-WinSCPOption @config.Prod.Host
```

Today, I would have to store both of those configurations into another variable first to use them:
```PowerShell
$qaConfig = $config.QA.Host
$prodConfig = $config.Prod.Host

New-WinSCPOption @qaConfig
New-WinSCPOption @prodConfig
```

It just adds more work when I could have just splat directly to them.

**Comment by adityapatwardhan** [comment]
@PowerShell/powershell-committee  Please re-review as the author has responded.

**Comment by joeyaiello** [comment]
@dragonwolf83 we in the @PowerShell/powershell-committee don't really understand why, in the 1st example where you set a variable for the params, is unclear

Unfortunately, we didn't have @BrucePay today in our @PowerShell/powershell-committee meeting, would love to get his take as well: as @JamesWTruher raised, this is all very tied to the parameter binder, and he may have an opinion there. 

Similarly, @KirkMunro, I'm really not seeing why the 2nd example is fundamentally more clear or concise to justify the additional complexity. I won't pretend to understand exactly in what way this is enabling future hashtable manipulation features, but we should establish first that those features are useful before leading with an implementation that we're not sure is a pre-requisite. You may have posted about those somewhere we haven't yet gotten to, but the value that said features would enable should be the front of the pipeline.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we are ok with the intent.  We ask that the tests be increased to cover multiple values rather than splatting a single value.

**Comment by KirkMunro** [comment]
Closing. See https://github.com/PowerShell/PowerShell/pull/10238#issuecomment-826091414.

**Pull request closed without merging**